main/src/library/LazyList.flix

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2020 Esben Bjerre
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace LazyList {
+
+    ///
+    /// An immutable lazy list.
+    ///
+    /// A lazy list is either the empty list represented by `Empty`, or
+    /// an element `x` followed by a lazy list `xs`.
+    /// Note that a lazy list is lazy on its tail but strict on its head.
+    /// The empty list is always strict.
+    ///
+    pub enum LazyList[t] {
+        case Empty,
+        case Cons(t, Lazy[LazyList[t]])
+    }
+
+    pub def empty(): LazyList[a] = Empty
+
+    pub def isEmpty(xs: LazyList[a]): Bool = xs == Empty
+
+    pub def cons(x: a, xs: LazyList[a]): LazyList[a] =
+        Cons(x, lazy xs)
+
+    pub def repeat(x: a): LazyList[a] =
+        unfold(_ -> Some(x, x), x)
+
+    pub def head(xs: LazyList[a]): Option[a] = match xs {
+        case Empty => None
+        case Cons(h, _) => Some(h)
+    }
+
+///    TODO: Unable to unify the types: 'Lazy' and 'LazyList'.
+///    This seems to be caused by the compiler transforming
+///    x :: acc to Cons(x, acc) which calls the Cons constructor in
+///    LazyList instead of List
+///
+///    pub def toList(xs: LazyList[a]): List[a] =
+///        foldRight((x, acc) -> x :: acc, Nil, xs)
+
+    pub def append(xs: LazyList[a], ys: LazyList[a]): LazyList[a] =
+        foldRight((x, acc) -> cons(x, acc), ys, xs)
+
+    pub def fold(f: (b, a) -> b & e, s: b, xs: LazyList[a]): b & e = foldLeft(f, s, xs)
+
+    pub def foldLeft(f: (b, a) -> b & e, s: b, xs: LazyList[a]): b & e = match xs {
+        case Empty => s
+        case Cons(h, t) => foldLeft(f, f(s, h), force t)
+    }
+
+    pub def foldRight(f: (a, b) -> b & e, s: b, xs: LazyList[a]): b & e = match xs {
```

Comment:
Question: Will this not force the entire list? (and hence e.g. exists would force the entire list, which should be unnecessary?)

Reply:
I think you're right. I took inspiration from Functional Programming in Scala, but I missed that they do the 2nd argument to `f` by-name.

Reply:
I'm a bit stuck with this method. To prevent forcing the entire list it is possible to define it as seen below. But I do not think this is the type signature we want.
```
pub def foldRight(f: (a, () -> b) -> b, s: () -> b, xs: LazyList[a]): b = match xs {
    case Empty => s()
    case Cons(h, t) => f(h, () -> foldRight(f, s, force t))
}
```

Reply:
After thinking some more I'm actually not sure the above works.

Reply:
I think *I* was confused. This stuff is also new to me.

A fold right of `1 :: 2 :: 3 :: Nil` looks like `f(1, f(2, f(3, ...)))`. This means, the first thing to do is to apply `f` to the last element of the list. This requires the entire list to be forced.

Hmm but maybe not:

http://voidmainargs.blogspot.com/2011/08/folding-stream-with-scala.html

Now I am confused on a higher level.

Reply:
https://stackoverflow.com/questions/48654292/why-does-foldright-work-for-infinite-lists